### PR TITLE
ci: [M12] CONF-2 follow-up: the macOS CI job is now 37m48s against a 45m timeout

### LINKED ISSUE #315: [M12] CONF-2 follow-up: the macOS CI job is now 37m48s against a 45m timeout
Follow-up from #303 (PR #314, merged 2026-08-20). Filed by the `/backlog` run that merged it,
because the capability landed but one of the PR's own Definition-of-done criteria did not hold.

## What #303 achieved (not in dispute)

CONF-2 is genuinely fixed. The five cross-backend comparisons in `tests/test_backend_parity.py`
that were guarded by `@pytest.mark.skipif(not (HAVE_MLX and HAVE_TORCH))` now **execute** on the
macOS job. Read from the run log rather than the checkmark, per that PR's criterion 2:

```
full suite + smoke gate + cross-backend parity (macOS)
  Cross-backend parity — MLX vs CUDA/torch-CPU must RUN here (#303)
  8 passed in 6.29s
```

No `SKIPPED` line names any of the five. The parity tests themselves cost **6.29 seconds**.

## The criterion that failed

#303's criterion 5 read:

> **`full-macos` runtime stays acceptable and the job count is unchanged.** Wall clock ≤ ~15 min
> (baseline 9m16s; `timeout-minutes: 45`)

Measured on PR #314's own CI run (`32337001751`):

| | value |
|---|---|
| baseline before #303 | 9m16s |
| criterion 5 ceiling | ~15 min |
| **actual** | **37m48s** (`05:48:39Z` → `06:26:27Z`) |
| `timeout-minutes` | 45 |
| **headroom** | **~7 min** |

The job passed, so CI was green and the PR merged legitimately. But the criterion was not met, and
nothing in the merge path observed that — which is the specific failure mode `docs/feature-matrix.md`
exists to prevent.

## Where the time actually goes

Not the parity tests. The main-suite step is the cost:

```
pytest -q -rs --ignore=tests/test_backend_parity.py
  1577 passed, 90 skipped, 15 warnings in 2165.37s (0:36:05)
```

Installing CPU `torch` alongside `mlx` un-skips a large set of tests that previously never ran on
this runner, so some increase is the *intended* consequence of #303 — more is genuinely being
verified. The question this issue asks is whether **4x the baseline** is the right price, and what
happens when the suite grows into the remaining 7 minutes.

## Why it matters

A job 7 minutes from its own timeout is a latent CI failure, and its failure mode is bad: an
`ETIMEDOUT` on the one job that carries the MLX↔CUDA parity guarantee reads as infrastructure
flake, not as a parity regression. The guard would go dark in a way that looks like noise — the
same shape as the original CONF-2 defect, where a green job hid five silent skips.

## Options (not yet decided)

1. **Raise `timeout-minutes`** — cheapest, buys headroom, addresses none of the cost.
2. **Split the macOS job in two** — a fast job carrying the parity step (6.29s) plus the smoke gate,
   and a slower full-suite job. Keeps the parity guarantee on a job that cannot be starved by
   unrelated suite growth. Mirrors the split #302 just made for the same reason.
3. **Move the full macOS suite to `scheduled-parity.yml`** and keep only parity + smoke on PR CI.
   Cuts PR cost most, but weakens per-PR coverage — and would need the same explicit "a green PR
   does not prove X" note the poc gate already carries.
4. **Profile first.** 9m16s → 36m for +~280 tests is a suspiciously large jump; it may be a handful
   of slow newly-un-skipped tests rather than a broad shift. Worth measuring with `--durations=25`
   before choosing 1–3.

Option 4 then 2 is the suggested order, but this is a decision, not a defect with one fix.

## Definition of done

- [ ] `--durations=25` output for the macOS full suite recorded here, identifying what the added
      time actually is
- [ ] A decision among the options above, written down with its trade-off
- [ ] macOS PR-path wall clock back under a stated ceiling, with that ceiling asserted mechanically
      rather than described — so this cannot silently regress again
- [ ] The five parity comparisons still provably RUN, verified from log text (`N passed`, no
      `SKIPPED` naming them), not from a green check
- [ ] `docs/feature-matrix.md` CONF-2 and `docs/test-plan.md` updated in the same commit as the code

### PR DESCRIPTION
**Pull request by travisgalloway**
Closes #315

## Summary

#303 gave `full-macos` both backends so the five MLX↔torch comparisons would actually execute. They did — and the same install un-skipped a large torch-gated set, taking the job from **9m16s → 37m48s** against its own `timeout-minutes: 45`. Nothing in the merge path observed that, because a job seven minutes short of its timeout still shows a green check.

The failure that sets up is specific: when the suite crosses 45 minutes GitHub kills the job with no explanation, and an unexplained kill on the *one* job carrying the MLX↔CUDA parity guarantee reads as infrastructure flake rather than as a parity regression — CONF-2's original failure shape wearing a different hat. `timeout-minutes` is a stop, not a signal.

- **Split the job** (issue option 2). The parity step and the MLX smoke gate cost **14 s combined**; the suite costs ~35 min. They now live on a new `parity-macos` job (~2 min against `timeout-minutes: 15`), so suite growth can no longer starve the gate. `full-macos` keeps its id and is now only the suite — including the `cuda` extra, because `test_upcycle.py` / `test_parallel.py` / `test_bench_config_export.py` / `test_bfcl_adapter.py` are torch-gated and fall outside `cuda-cpu`'s glob.
- **Gave the suite step a budget it measures itself against** (`MACOS_SUITE_BUDGET_SECONDS`), failing with an `::error::` that names #315 and the remedy. A failing suite stays a failing suite — never "over budget".
- **`tests/test_ci_macos_budget.py`** (new, portable) pins the budget literal, both `timeout-minutes`, that the budget is actually *read* by a step, and `budget + 300s <= timeout` so the legible guard always fires before `ETIMEDOUT`.
- Contracts and docs updated in the same commit: `PR_PUSH_JOBS` 8 → 9, `DESIGNATED_JOB` → `parity-macos`, `CLAUDE.md`, `docs/design/03-conformance.md`, `docs/test-plan.md` (DATA-1 / TOK-4 / MODEL-2 / TRAIN-1 / CONF-2 / OPS-1), `docs/feature-matrix.md` CONF-2, `docs/local-development.md`.

`full-macos` deliberately stays at `timeout-minutes: 45` in the first commit: tightening it before the runner's own `--durations=25` profile exists would `ETIMEDOUT` the very run that has to print the table. The profile from this PR's first run is what sets the final budget.

## Testing
- [x] Tests added/updated — `tests/test_ci_macos_budget.py` (8 cases)
- [x] All tests passing — `pytest -q -rs`: **1772 passed, 22 skipped**; `MONICA_REQUIRE_BOTH_BACKENDS=1 pytest tests/test_backend_parity.py`: **8 passed**
- [ ] Manual testing completed — the macOS wall clock can only be measured on the runner; read from this PR's run

### CODE REVIEW

File: docs/test-plan.md [hunk 1]
```diff
@@ -151,6 +151,6 @@ Last audited 2026-08-18 (`/closure-audit`, whole repo: 81 capabilities × 131 te
 
 | ID | Unit | Integration | E2E | Edge cases covered | Gaps |
 |----|------|-------------|-----|--------------------|------|
-| OPS-1 | n/a | n/a | `scripts/smoke_test.py`, CI `smoke-linux` + `full-macos` | resume exactness + eval | does not cover `train.py`'s stream-resume (see TRAIN-2) |
+| OPS-1 | n/a | n/a | `scripts/smoke_test.py`, CI `smoke-linux` + `parity-macos` | resume exactness + eval | does not cover `train.py`'s stream-resume (see TRAIN-2) |
 | OPS-2 | `test_workflow_triggers.py` | n/a | `ci.yml` (8 jobs, PR/push + monthly schedule) + `scheduled-parity.yml` (4 jobs, dispatch/weekly schedule) | portable/seam guard, both backends, Swift parity; the trigger×job matrix itself; `ci.yml`'s cron literal `"43 9 3 * *"` and its event-scoped `concurrency.group` (#312) | 4 of 12 jobs are in a workflow with no PR/push trigger (see ENGINE-6); the monthly cadence bounds drift-detection latency to ~1 month, and nothing alerts on a red scheduled run beyond GitHub's default notification |
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply docs/test-plan.md]: ✅ Fixed in 6ce5fd2 — OPS-2 row now says ci.yml (9 jobs) and 4 of 13 total, matching tests/test_workflow_triggers.py.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -179,6 +196,67 @@ jobs:
...
+        # no explanation, and an unexplained kill reads as flake. This measures its own wall
+        # clock and fails with an ::error:: naming the cause and the remedy. Note the
+        # set +e / STATUS dance — a FAILING SUITE MUST STAY A FAILING SUITE, never be
+        # reported as "over budget", and the elapsed line prints on both paths.
+        env:
+          MACOS_SUITE_BUDGET_SECONDS: "1200"
+        run: |
+          START=$SECONDS
+          set +e
+          pytest -q -rs --ignore=tests/test_backend_parity.py --durations=25
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply .github/workflows/ci.yml:222]: ✅ Fixed in 6ce5fd2 — raised MACOS_SUITE_BUDGET_SECONDS to 2400s (recorded 37m48s baseline + buffer, capped by the 300s margin below the 45-min timeout) in both ci.yml and tests/test_ci_macos_budget.py. Will tighten toward 20 min once the suite itself is faster.

**Comment by travisgalloway** [comment]
Addressed review feedback:

- @copilot-pull-request-reviewer: budget/timeout mismatch (MACOS_SUITE_BUDGET_SECONDS 1200s vs the ~35-min suite baseline) fixed in 6ce5fd2 — budget raised to 2400s in both ci.yml and tests/test_ci_macos_budget.py.
